File: android/src/main/kotlin/map/naver/plugin/net/note11/naver_map_plugin/NaverPathsController.kt
package map.naver.plugin.net.note11.naver_map_plugin

import map.naver.plugin.net.note11.naver_map_plugin.Convert.toCoords
import map.naver.plugin.net.note11.naver_map_plugin.Convert.toColorInt
import map.naver.plugin.net.note11.naver_map_plugin.Convert.toOverlayImage
import com.naver.maps.map.NaverMap
import com.naver.maps.map.overlay.Overlay
import com.naver.maps.map.overlay.PathOverlay
import java.util.HashMap
import kotlin.math.roundToInt

class NaverPathsController(
    private val naverMap: NaverMap,
    private val onClickListener: Overlay.OnClickListener,
    private val density: Float
) {
    private val idToController = HashMap<String?, PathController>()
    fun set(pathData: List<*>?) {
        if (pathData == null || pathData.isEmpty()) return
        for (json in pathData) {
            val data = json as HashMap<String, Any>
            val id = data["pathOverlayId"] as String?
            if (idToController.containsKey(id)) {
                idToController[id]!!.interpret(data)
            } else {
                val controller = PathController(data)
                controller.setOnClickListener(onClickListener)
                idToController[controller.id] = controller
                controller.setMap(naverMap)
            }
        }
    }

    fun remove(pathIds: List<*>?) {
        if (pathIds == null || pathIds.isEmpty()) return
        for (json in pathIds) {
            val idToRemove = json as String?
            val pathToRemove = idToController[idToRemove]
            pathToRemove!!.setMap(null)
            pathToRemove.setOnClickListener(null)
            idToController.remove(idToRemove)
        }
    }

    inner class PathController(option: HashMap<String, Any>) {
        val id: String = (option["pathOverlayId"] as String?)!!
        private val path: PathOverlay = PathOverlay()
        fun interpret(option: HashMap<String, Any>) {
            val latLngData = option["coords"] as List<*>?
            if (latLngData != null && latLngData.size >= 2) path.coords = latLngData.toCoords()
            val globalZIndex = option["globalZIndex"]
            if (globalZIndex != null) path.globalZIndex = globalZIndex as Int
            val hideCollidedCaptions = option["hideCollidedCaptions"]
            if (hideCollidedCaptions != null) path.isHideCollidedCaptions =
                hideCollidedCaptions as Boolean
            val hideCollidedMarkers = option["hideCollidedMarkers"]
            if (hideCollidedMarkers != null) path.isHideCollidedMarkers =
                hideCollidedMarkers as Boolean
            val hideCollidedSymbols = option["hideCollidedSymbols"]
            if (hideCollidedSymbols != null) path.isHideCollidedSymbols =
                hideCollidedSymbols as Boolean
            val color = option["color"]
            if (color != null) path.color = toColorInt(color)
            val outlineColor = option["outlineColor"]
            if (outlineColor != null) path.outlineColor =
                toColorInt(outlineColor)
            val passedColor = option["passedColor"]
            if (passedColor != null) path.passedColor = toColorInt(passedColor)
            val passedOutlineColor = option["passedOutlineColor"]
            if (passedOutlineColor != null) path.passedOutlineColor =
                toColorInt(passedOutlineColor)
            val patternImage = option["patternImage"]
            if (patternImage != null) path.patternImage =
                toOverlayImage(patternImage)
            val patternInterval = option["patternInterval"]
            if (patternInterval != null) path.patternInterval =
                (patternInterval as Int * density).roundToInt()
            val progress = option["progress"]
            if (progress != null) path.progress = progress as Double
            val width = option["width"]
            if (width != null) path.width = (width as Int * density).roundToInt()
            val outlineWidth = option["outlineWidth"]
            if (outlineWidth != null) path.outlineWidth =
                (outlineWidth as Int * density).roundToInt()
        }

        fun setMap(naverMap: NaverMap?) {
            path.map = naverMap
        }

        fun setOnClickListener(listener: Overlay.OnClickListener?) {
            path.onClickListener = listener
        }

        init {
            path.tag = this
            interpret(option)
        }
    }
}

File: ios/Classes/src/NaverPathsController.swift
//
//  NaverPathsController.swift
//  naver_map_plugin
//
//  Created by Maximilian on 2020/08/28.
//

import Foundation
import Flutter
import NMapsMap

class NPathController: NSObject {
    let path:NMFPath
    let id: String
    let registrar: FlutterPluginRegistrar
    
    init(json: NSDictionary, registrar: FlutterPluginRegistrar) {
        assert(json["pathOverlayId"] != nil && json["coords"] != nil)
        path = NMFPath()
        self.id = json["pathOverlayId"] as! String
        self.registrar = registrar
        
        super.init()
        path.userInfo = ["path" : self]
        
        interprete(json)
    }// init
    
    func interprete(_ json: NSDictionary){
        if let coordData = json["coords"] as? Array<Any>, coordData.count > 1 {
            var coords: Array<NMGLatLng> = []
            coordData.forEach { (latLngData) in
                coords.append(toLatLng(json: latLngData))
            }
            path.path = NMGLineString(points: coords)
        }
        if let globalZIndex = json["globalZIndex"] as? Int {
            path.globalZIndex = globalZIndex
        }
        if let hideCollidedCaptions = json["hideCollidedCaptions"] as? Bool{
            path.isHideCollidedCaptions = hideCollidedCaptions
        }
        if let hideCollidedMarkers = json["hideCollidedMarkers"] as? Bool{
            path.isHideCollidedMarkers = hideCollidedMarkers
        }
        if let hideCollidedSymbols = json["hideCollidedSymbols"] as? Bool{
            path.isHideCollidedSymbols = hideCollidedSymbols
        }
        if let color = json["color"] as? NSNumber {
            path.color = toColor(colorNumber: color)
        }
        if let outlineColor = json["outlineColor"] as? NSNumber{
            path.outlineColor = toColor(colorNumber: outlineColor)
        }
        if let outlineWidth = json["outlineWidth"] as? CGFloat {
            path.outlineWidth = outlineWidth
        }
        if let passedColors = json["passedColor"] as? NSNumber{
            path.passedColor = toColor(colorNumber: passedColors)
        }
        if let passedOutlineColor = json["passedOutlineColor"] as? NSNumber{
            path.passedOutlineColor = toColor(colorNumber: passedOutlineColor)
        }
        if let patternImage = json["patternImage"] as? String {
            path.patternIcon = toOverlayImage(assetName: patternImage, registrar: registrar)
        }
        if let patternImageFromFile = json["patternImageFromPath"] as? String {
            path.patternIcon = toOverlayImageFromFile(imagePath: patternImageFromFile)
        }
        if let patternInterval = json["patternInterval"] as? UInt {
            path.patternInterval = patternInterval
        }
        if let progress = json["progress"] as? Double{
            path.progress = progress
        }
        if let width = json["width"] as? CGFloat {
            path.width = width
        }
    }
    
    func setMap(_ map:NMFNaverMapView?){
        path.mapView = map?.mapView
    }
}

class NaverPathController:NSObject {
    private var idToController = Dictionary<String, NPathController>()
    let registrar: FlutterPluginRegistrar
    let naverMap: NMFNaverMapView
    let touchHandler: (NMFOverlay)->Bool
    
    init(naverMap: NMFNaverMapView, registrar: FlutterPluginRegistrar, touchHandler: @escaping (NMFOverlay)->Bool) {
        self.naverMap = naverMap
        self.registrar = registrar
        self.touchHandler = touchHandler
    }
    
    func set(jsonArray: Array<Any>) {
        jsonArray.forEach { (json) in
            DispatchQueue.main.async {
                if let data = json as? NSDictionary {
                    if let id = data["pathOverlayId"] as? String, self.idToController[id] != nil {
                        // 객체 update.. 속성값 변경
                        self.idToController[id]?.interprete(data)
                        return
                    }
                    let pathToAdd = NPathController(json: data, registrar: self.registrar)
                    pathToAdd.setMap(self.naverMap)
                    pathToAdd.path.touchHandler = self.touchHandler
                    self.idToController[pathToAdd.id] = pathToAdd
                }
            }
        }
    }
    
    func remove(jsonArray: Array<Any>) {
        jsonArray.forEach { (rawId) in
            DispatchQueue.main.async {
                if let id = rawId as? String, let pathToRemove = self.idToController[id] {
                    pathToRemove.path.touchHandler = nil
                    pathToRemove.setMap(nil)
                    self.idToController.removeValue(forKey: id)
                }
            }
        }
    }
    
}
